### ISSUE TITLE
torch.multinomial with replacement=True produces inaccurate results for large number of categories

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#43115]
## 🐛 Bug

`torch.multinomial` with `replacement=True` produces very inaccurate results when number of categories is large (and/or probabilities have very different values)

## To Reproduce
One of the ways to reproduce (there could be multiple ways to construct adversarial examples):
```
import torch
for ncat in (2**8, 2**22, 2**24):
    probs=torch.empty(ncat) 
    half = ncat//2    
    probs[half:]=1
    probs[:half-1]=2
    out=torch.multinomial(probs, num_samples=10**5, replacement=True) 
    print("number of categories {:10}, number of samples in upper half {:8} ".format(ncat, (out>=half).sum().item())) #would expect 1/3 (~33333) of generated values to be >=half, true for ncat 2**8, 2**22, 0 for ncat 2**24
```
## Additional context

This is due to how multinomial sampling with replacement is implemented - we compute cumsum of the probs values, then normalize the sum to 1 and compute boundaries as cumsum(probs)/sum(probs). If there are many categories (or particularly small probabilities) boundaries will collapse to be the same value. 
We disallow more than 2**24 categories for that reason, but large inaccuracies can start before that. 
Related: #2576
